Autoverify fix -  MissingAutoVerifyDetector final cleanup

olivandcode opened the pull request:
Addressing the comment in https://github.com/google/android-security-lints/pull/12#issuecomment-3527813294

Commit:
Fix the logic for CustomSchemeDetector to comply with the documentation.

Commit:
Refactor CustomSchemeDetector to MissingAutoVerifyDetector

- Renamed class to `MissingAutoVerifyDetector` to better reflect the specific check being performed.
- Added comprehensive KDoc.
- Refactored XML parsing to use idiomatic Kotlin `for-in` loops by adding a `NodeList.iterator()` extension function.
- Consolidated duplicate logic into a reusable `hasCategory()` helper function.
- Updated Issue ID and description to match the new class name.

Commit:
Refactor/Docs: Complete MissingAutoVerifyDetector cleanup and documentation

* Fix remaining references to the old detector name (`CustomSchemeDetector`).
* Clarify the check's purpose in the README.
* Update sample application's `AndroidManifest.xml`.

bjiang7 commented on README.md:
```diff
@@ -76,7 +76,7 @@ on [![Google's Maven Repository](https://img.shields.io/maven-metadata/v?metadat
 | `TapjackingVulnerable`               | [`TapjackingDetector`](checks/src/main/java/com/example/lint/checks/TapjackingDetector.kt)             | Views without the `filterTouchesWhenObscured` attribute are susceptible to tapjacking attacks by other apps obscuring the UI to trick the user into performing certain actions.                       |
 | `StrandhoggVulnerable`               | [`StrandhoggDetector`](checks/src/main/java/com/example/lint/checks/StrandhoggDetector.kt)             | Android previously had a bug in task reparenting in earlier versions, allowing malicious applications to hijack legitimate user actions and trick users into providing credentials to malicious apps. |
-| `MissingAutoVerifyAttribute`         | [`CustomSchemeDetector`](checks/src/main/java/com/example/lint/checks/CustomSchemeDetector.kt)         | Exporting sensitive functionality via custom URL schemes should be properly protected by making use of app link verification.                                                                         |
+| `MissingAutoVerifyAttribute`         | [`MissingAutoVerifyDetector`](checks/src/main/java/com/example/lint/checks/MissingAutoVerifyDetector.kt)         | Failing to implement App Links verification leaves the app susceptible to the Insecure Deep Links risk, potentially allowing other applications to hijack the domain's web URLs and intercept traffic intended for the legitimate app.              ...
```
Uber nit: "susceptible to the risk of Insecure Deep Links"

olivandcode replied:
Fixed in the next commit

bjiang7 commented on checks/src/main/java/com/example/lint/checks/MissingAutoVerifyDetector.kt:
```diff
@@ -0,0 +1,141 @@
+
+    override fun visitElement(context: XmlContext, element: Element) {
+        if (!hasViewAction(element) ||
+            !hasCategory(element, "android.intent.category.BROWSABLE") ||
```
Is it worth adding "android.intent.category.BROWSABLE" and "android.intent.category.DEFAULT" as constants in the companion object of the class?

olivandcode replied:
Fixed in the next commit

bjiang7 commented on checks/src/main/java/com/example/lint/checks/MissingAutoVerifyDetector.kt:
```diff
@@ -0,0 +1,141 @@
+        val actions = element.getElementsByTagName(TAG_ACTION)
+
+        for (action in actions) {
+            if (action.getAttributeNS(ANDROID_URI, "name") == "android.intent.action.VIEW") {
```
Same here, worth making this into a constant in the companion object?

olivandcode replied:
Fixed in the next commit

bjiang7 commented on checks/src/main/java/com/example/lint/checks/MissingAutoVerifyDetector.kt:
```diff
@@ -0,0 +1,141 @@
+    companion object {
+        private const val EXPLANATION = """
+            Intent filters that handle `http` or `https` schemes and include \
+            `android.intent.action.VIEW`, `android.intent.category.BROWSABLE`, and \
```
If you made these into constants, then you could make EXPLANATION a format string that uses these constants.

https://kotlinlang.org/docs/strings.html#string-templates

olivandcode replied:
Fixed in the next commit

bjiang7 commented on checks/src/test/java/com/example/lint/checks/MissingAutoVerifyDetectorTest.kt:
```diff
@@ -65,7 +67,7 @@ class CustomSchemeDetectorTest : LintDetectorTest() {
 
     @Test
-    fun testWhenNoAutoVerifyAttributeSpecifiedOnRegularSchemeIntentFilter_noWarning() {
+    fun testWhenHttpSchemeButMissingCategories_noWarning() {
```
Uber nit: showsNoWarning

Please fix throughout.

olivandcode replied:
Fixed in the next commit

bjiang7 reviewed (changes requested):
LGTM Olivier, just a few really small nits. I'll merge your changes once they're addressed :)

Commit:
Improve MissingAutoVerifyDetector constants and test names

Improves the consistency and maintainability of the App Links lint detector:
- Extracts `ACTION_VIEW`, `CATEGORY_BROWSABLE`, `CATEGORY_DEFAULT`, and HTTP/S schemes into constants within `MissingAutoVerifyDetector.companion object`.
- Updates the `EXPLANATION` string to use string interpolation with these new constants.
- Renames all test methods in `MissingAutoVerifyDetectorTest` to follow the standard `testWhen[Condition]_shows[Warning|NoWarning]` pattern (e.g., `testWhenHttpSchemeButMissingCategories_showsNoWarning`).

bjiang7 committed:
Merge branch 'main' into autoverify-fix

matejdro commented:
Any chance we can get a release with this fixed?

bjiang7 commented:
Version 1.0.4 should include this fix (released in mid-Dec), let us know if you're still not seeing the fix in this version.

Just a heads up that a migration is happening internally on how releases are published, so there will be a delay if a new release is needed. We're doing our best, but are now maintaining this library in our spare time as we've been moved onto other projects, so appreciate your patience and understanding on this as we find a longer term solution :)

matejdro commented:
Ah, it appears to be fixed, I got confused by the fact that the 1.0.4 commit was made before merging of this. Thanks!

bjiang7 commented:
No worries, thanks for confirming it's fixed in the latest version!
